llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp

```diff
@@ -3470,17 +3470,19 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
             auto *MIBMD = cast<const MDNode>(MDOp);
             MDNode *StackMDNode = getMIBStackNode(MIBMD);
             assert(StackMDNode);
-            SmallVector<unsigned> StackIdsFromMetadata;
             CallStack<MDNode, MDNode::op_iterator> StackContext(StackMDNode);
-            for (auto ContextIter =
-                     StackContext.beginAfterSharedPrefix(CallsiteContext);
+            auto ContextIterBegin =
+                StackContext.beginAfterSharedPrefix(CallsiteContext);
+            // Skip the checking on the first iteration.
+            uint64_t LastStackContextId = *ContextIterBegin == 0 ? 1 : 0;
```

Comment:
I ran into an assert in the operator* method because ContextIterBegin was already at StackContext.end() (after skipping the shared prefix above). I think this should be:
```
uint64_t LastStackContextId = (ContextIter != StackContext.end() && *ContextIterBegin == 0) ? 1 : 0;
```
Can you send a fix?

Reply:
Ok, OOO, fix later.

Reply:
I think ContextIter should't at StackContext.end() because we already have assert `assert(AI != FS->allocs().end());`. How about adding one more assert `assert(ContextIterBegin != StackContext.end());`? https://github.com/llvm/llvm-project/pull/81004

Reply:
The allocs list on the FS is different than the list of context ids on a given MIB though, so not sure how that assert is related?

The assert you propose will fire in the case that I saw, this is the reason we are getting the assert in operator*. Since this is breaking our memprof builds I have a fix and a testcase that I will send for review shortly. I'm not exactly sure why we are seeing this case, I assume it happened during inlining but in that case we should have ended up replacing the memprof metadata with an attribute. I have a FIXME in the test to investigate this. But at this point I want to unblock our builds.

Reply:
https://github.com/llvm/llvm-project/pull/81008

Reply:
>  But at this point I want to unblock our builds.

I think ContextIter should't at StackContext.end(). So maybe add one assert is more reasonable. It's ok to do this to unblock your builds. Thanks for reply.